src/hotspot/os/windows/os_windows.cpp

```diff
@@ -2150,7 +2150,7 @@ size_t os::lasterror(char* buf, size_t len) {
     size_t n = strlen(s);
     if (n >= len) n = len - 1;
     strncpy(buf, s, n);
-    buf[n] = '\0';
+    buf[n + 1] = '\0';
```

Comment:
Unless os::lasterror's len parameter is not supposed to include space for the null terminator, but I couldn't find any documentation or comment that says so. But if that were the case, then this whole code segment would be wrong, not just my change

Ahh, the joys of working with Windows error handling :)

Reply:
Nevermind, just checked the POSIX version and it also has this flaw, so it's not just Windows

Reply:
We have a buffer `buf` of length `len` into which we need to copy a string `s` of length `n` and nul-terminate it. So the string will occupy `buf[0]` to `buf[n-1]` with the nul at `buf[n]`.

If the string + nul is too big for the buffer `n >= len` then we set `n = len - 1` so we copy a truncated `s` and then store the nul at `buf[n]`.

There is no bug here.

Reply:
I don't quite follow, strlen only returns the length of the string _without_ the null terminating character, so the string's actual character data does indeed occupy buf[0] to buf[n] (since n = strlen(s)), meaning that the null terminator would be at n + 1. When strncpy is called later on it always reaches the count n before it reaches the null character, so the string that was copied is never null terminated, meaning that nulling buf[n] would actually erroneously discard the last character in the buffer

Reply:
The actual characters of a `n` length string occupy `buf[0]` to `buf[n-1]`, with the terminating nul at `buf[n]`.

Reply:
I understand that that would be the case if n = strlen(s) + 1, but here it's only n = strlen(s). Am I missing something?

Reply:
```
char* s = "A";
int n = strlen(s); // n == 1
char buf[2];
strcpy(buf, s, n)  // buf[0] == 'A'
buf[n] = '\0`;    // buf[1] == '0\`
```

Reply:
> I understand that that would be the case if n = strlen(s) + 1, but here it's only n = strlen(s). Am I missing something?

Array addressing via [] is zero based. See Davids example. For "ABC" A is index 0, C is 2, \0 is 3.

Reply:
*facepalm

I completely forgot about array indexing. Sorry for all the trouble